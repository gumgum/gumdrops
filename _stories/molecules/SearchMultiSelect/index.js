import React from 'react';
import { select, button, text, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import SearchMultiSelect from '../../../components/molecules/SearchMultiSelect';
import FormGroupLabel from '../../../components/atoms/FormGroupLabel';
import FormGroup from '../../../components/molecules/FormGroup';

class TestSearchMultiSelect extends React.Component {
    state = {
        names: namesList
    };

    componentDidMount() {
        this._setFocus();
    }

    _setFocus = () => {
        if (this._inputRef.current) {
            this._inputRef.current.focus();
        }
    };

    _inputRef = React.createRef();

    _updateNames = names => {
        this.setState({ names });
        action('SearchMultiSelect Updated')(names);
    };

    _handleInputChange = value => action('SearchMultiSelect current value')(value);

    render() {
        return (
            <FormGroup>
                <FormGroupLabel text="names" />
                {button('Set focus', this._setFocus)}
                <SearchMultiSelect
                    inputRef={this._inputRef}
                    placeholder={text('Placeholder', 'My placeholder')}
                    options={this.state.names}
                    update={this._updateNames}
                    onChange={this._handleInputChange}
                    context={select('Context', contextOptions, 'primary')}
                    size={select('Size', ['xs', 'sm', 'md'], 'md')}
                    multiTerm={boolean('Multiple Terms', false)}
                    searchKeys={boolean('Search Keys', false)}
                    termDivider={text('Term Divider')}
                />
            </FormGroup>
        );
    }
}

const component = () => <TestSearchMultiSelect />; //eslint-disable-line react/no-multi-comp

// Props data
const namesList = [
    {
        name: 'Jose Santos',
        isSelected: true,
        key: 0
    },
    {
        name: 'Michele Larson',
        isSelected: true,
        key: 1
    },
    {
        name: 'Serge Basile',
        isSelected: false,
        key: 2
    },
    {
        name: 'Eder "Quesarito" Sanchez',
        isSelected: false,
        key: 3
    },
    {
        name: 'David "Henry" Mejorado',
        isSelected: false,
        key: 4
    },
    {
        name: 'Mahdiye \\ Maddie Jamali',
        isSelected: false,
        key: 5
    }
];

const contextOptions = [
    'primary',
    'secondary',
    'success',
    'warning',
    'danger',
    'blue',
    'gold',
    'green',
    'orange',
    'purple',
    'red',
    'darkblue',
    'darkgreen',
    'darkred',
    'primary-dark',
    'success-dark',
    'warning-dark',
    'danger-dark',
    'blue-dark',
    'gold-dark',
    'green-dark',
    'orange-dark',
    'purple-dark',
    'red-dark',
    'darkblue-dark',
    'darkgold-dark',
    'darkgreen-dark',
    'darkred-dark'
];

TestSearchMultiSelect.displayName = 'SearchMultiSelect';

export default [readme, component];
