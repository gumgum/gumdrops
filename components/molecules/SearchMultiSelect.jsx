import React, { Component } from 'react';
import PropTypes from 'prop-types';
import arraysEqual from '../utils/arraysEqual';
import cx from 'classnames';
import Tag from '../atoms/Tag';

import escapeRegExp from '../utils/escapeRegExp';

class SearchMultiSelect extends Component {
    constructor({ options, searchKeys, multiTerm, termDivider, filter }) {
        super();
        const searchConfig = { searchKeys, multiTerm, termDivider };
        this.state = {
            isOpen: false,
            isTagsOpen: false,
            searchTerm: '',
            currentIndex: 0,
            // Indexes of options that match the input term
            matchingIndexes: filter(options, '', searchConfig)
        };
    }

    componentDidMount() {
        const el = this.container;
        el.addEventListener('keydown', this._handleKeyboardEvents);
        window.addEventListener('click', this._closeOnClickOutside);
    }

    UNSAFE_componentWillReceiveProps({ options, searchKeys, multiTerm, termDivider, filter }) {
        const old = this.props;
        if (!arraysEqual(options, old.options)) {
            const searchConfig = { searchKeys, multiTerm, termDivider };
            const matchingIndexes = filter(options, this.state.searchTerm, searchConfig);
            this.setState({ matchingIndexes });
        }
    }

    UNSAFE_componentWillUpdate(_, { searchTerm }) {
        const { options, searchKeys, multiTerm, termDivider, filter } = this.props;
        const searchConfig = { searchKeys, multiTerm, termDivider };
        const { searchTerm: oldTerm, currentIndex, isOpen } = this.state;

        // If the term is the same, or the component is not open, do nothing
        if (searchTerm === oldTerm || !isOpen) return;

        // Dictionary with indexes of all matching options
        const matchingIndexes = filter(options, searchTerm, searchConfig);

        // If the previous selected option is not in the matched list
        // move to the first matching
        const nextIndex = matchingIndexes.includes(currentIndex)
            ? currentIndex
            : matchingIndexes[0];

        this.setState({ currentIndex: nextIndex, matchingIndexes });
    }

    componentWillUnmount() {
        const el = this.container;
        el.removeEventListener('keydown', this._handleKeyboardEvents);
        window.removeEventListener('click', this._closeOnClickOutside);
    }

    _updateSearchTerm = ({ target }) => {
        const { value: searchTerm } = target;
        const { onChange } = this.props;
        const externalOnChange = () => onChange && onChange(searchTerm);
        this.setState(
            {
                isOpen: true,
                searchTerm
            },
            externalOnChange
        );
    };

    _openSelect = () => {
        // If open aready, nothing to update
        if (this.state.isOpen) return;
        this.setState({ isOpen: true });
    };

    _closeSelect = () => this.setState({ isOpen: false });

    _toggleSelect = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }));

    _toggleTags = () => this.setState(prevState => ({ isTagsOpen: !prevState.isTagsOpen }));

    _clearAll = e => {
        e.stopPropagation();
        const { options, update } = this.props;
        const newOptions = options.map(o => ({ ...o, isSelected: false }));
        update(newOptions);
    };

    _toggleOption = index => {
        const { options, update } = this.props;
        const { isSelected, ...oldOption } = options[index];
        const newOptions = options.slice();
        newOptions[index] = { ...oldOption, isSelected: !isSelected };
        update(newOptions);
        this.setState({ currentIndex: index });
    };

    _removeOption = index => {
        const { options, update } = this.props;
        const oldOption = options[index];
        const newOptions = options.slice();
        newOptions[index] = { ...oldOption, isSelected: false };
        const tagCount = newOptions.reduce(
            (acc, { isSelected }) => (isSelected ? acc + 1 : acc),
            0
        );
        if (tagCount < 1) this.setState({ isTagsOpen: false });
        update(newOptions);
    };

    _handleKeyboardEvents = ({ keyCode }) => {
        switch (keyCode) {
            case 13: // enter
                this._toggleOnEnter();
                break;
            case 27: // esc
                this._closeSelect();
                break;
            case 40: // down
                this._selectNextOption();
                break;
            case 38: // up
                this._selectPrevOption();
                break;
        }
    };

    _toggleOnEnter() {
        const { currentIndex, isOpen } = this.state;
        if (!isOpen) return; // do nothing if closed
        this._toggleOption(currentIndex);
    }

    _selectNextOption() {
        const { matchingIndexes, currentIndex, isOpen } = this.state;
        if (!isOpen) {
            this._openSelect();
            return;
        }
        const i = matchingIndexes.indexOf(currentIndex);
        const nextIndex =
            matchingIndexes.length > i + 1 ? matchingIndexes[i + 1] : matchingIndexes[0];
        this.setState({ currentIndex: nextIndex });
    }

    _selectPrevOption() {
        const { matchingIndexes, currentIndex, isOpen } = this.state;
        if (!isOpen) {
            this._openSelect();
            return;
        }
        const i = matchingIndexes.indexOf(currentIndex);
        const last = matchingIndexes.length - 1;
        const prevIndex = i - 1 >= 0 ? matchingIndexes[i - 1] : matchingIndexes[last];
        this.setState({ currentIndex: prevIndex });
    }

    _closeOnClickOutside = ({ target }) => {
        const el = this.container;
        if (!el.contains(target)) this._closeSelect();
    };

    _getContainer = ref => (this.container = ref);

    render() {
        const { isOpen, isTagsOpen, currentIndex, matchingIndexes } = this.state;
        const { inputRef, options, context, placeholder, size } = this.props;

        const numberSelected = options.filter(o => o.isSelected).length;

        const rootClass = cx('gds-search-select', {
            'gds-search-select--open': isOpen
        });

        const openTagsClass = isTagsOpen ? 'gds-search-select__tag-holder--bubble-active' : '';

        const tagSize = size === 'sm' ? 'xs' : 'sm';
        const tagStyle = size === 'sm' ? { top: '0.4rem' } : {};

        return (
            <div ref={this._getContainer} className={rootClass}>
                <div className="gds-search-select__control">
                    {numberSelected !== 0 && (
                        <Tag
                            className="gds-search-select__tag-indicator"
                            context={context}
                            hasOption
                            onClick={this._toggleTags}
                            onOptionClick={this._clearAll}
                            optionIcon="bt-times"
                            optionLabel="Clear all tags"
                            size={tagSize}
                            style={tagStyle}
                            text={`${numberSelected} Selected`}
                        />
                    )}
                    <input
                        ref={inputRef}
                        onFocus={this._openSelect}
                        onClick={this._openSelect}
                        onChange={this._updateSearchTerm}
                        type="text"
                        placeholder={placeholder}
                        className={`gds-search-select__input gds-search-select__input--${size} ${
                            numberSelected > 0 ? hasTags : ''
                        }`}
                    />
                    <button
                        className="gds-search-select__toggle-button -cursor--pointer"
                        type="button"
                        onClick={this._toggleSelect}
                    />
                </div>
                <div className={`${TagHolderClasses} ${openTagsClass}`}>
                    <div className="gds-search-select__tag-overflow">
                        {options.map(({ name, key, isSelected }, index) => {
                            if (!isSelected) return;
                            return (
                                <Tag
                                    key={key}
                                    className="-m-a-1"
                                    context={context}
                                    hasOption
                                    onOptionClick={() => this._removeOption(index)}
                                    optionIcon="bt-times"
                                    optionLabel="Remove option"
                                    size="sm"
                                    text={name}
                                />
                            );
                        })}
                    </div>
                </div>
                <div className="gds-search-select__menu">
                    <div className="gds-search-select__menu-items">
                        {matchingIndexes.map(index => {
                            const { name, key, isSelected } = options[index];

                            const itemClass = cx('gds-search-select__menu-item', {
                                'gds-search-select__menu-item--selected': index === currentIndex
                            });

                            return (
                                <div
                                    key={key}
                                    onClick={() => this._toggleOption(index)}
                                    className={itemClass}>
                                    <label className="gds-search-select__checkbox">
                                        <input
                                            name={name}
                                            className="gds-search-select__checkbox-input"
                                            type="checkbox"
                                            checked={isSelected}
                                            readOnly
                                        />
                                        <span className="gds-search-select__checkbox-indicator" />
                                    </label>
                                    {name}
                                </div>
                            );
                        })}
                    </div>
                </div>
            </div>
        );
    }
}

const TagHolderClasses = `gds-search-select__tag-holder
    gds-search-select__tag-holder--bubble`;

const hasTags = 'gds-search-select__input--has-tag';

// Given an array of options, it returns the indexes where
// the 'name' property matches the 'term' regEx
const getMatchingIndexes = (options, query, config) => {
    const { searchKeys, multiTerm, termDivider } = config;
    const terms = multiTerm ? query.split(termDivider) : [query];
    const findMatches = q =>
        terms.some(term => {
            const escapedTerm = escapeRegExp(term);
            const regEx = new RegExp(escapedTerm, 'ig');
            return regEx.test(q);
        });
    return options.reduce((acc, { name, key }, index) => {
        const isNameMatch = findMatches(name);
        const isKeyMatch = searchKeys && findMatches(key);
        if (isNameMatch || isKeyMatch) acc.push(index);
        return acc;
    }, []);
};

SearchMultiSelect.displayName = 'SearchMultiSelect';

SearchMultiSelect.defaultProps = {
    context: 'primary',
    placeholder: '',
    size: 'md',
    filter: getMatchingIndexes,
    searchKeys: false,
    multiTerm: false,
    termDivider: /[ ,]+/
};

SearchMultiSelect.propTypes = {
    options: PropTypes.array.isRequired,
    update: PropTypes.func.isRequired,
    onChange: PropTypes.func,
    filter: PropTypes.func,
    context: PropTypes.string,
    placeholder: PropTypes.string,
    size: PropTypes.oneOf(['sm', 'md']),
    searchKeys: PropTypes.bool,
    multiTerm: PropTypes.bool,
    termDivider: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(RegExp)]),
    inputRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
};

export default SearchMultiSelect;
