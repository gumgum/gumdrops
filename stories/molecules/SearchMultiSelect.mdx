import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';

import * as SearchMultiSelectStories from './SearchMultiSelect.stories';
import SearchMultiSelect from '../../components/molecules/SearchMultiSelect';

# SearchMultiSelect

The `<SearchMultiSelect>` is similar to the `<MultiSelect>` component, but it is tailored for situations where there are way too many options that the `<MultiSelect>` becomes impractical.

As the name suggests, the `<SearchMultiSelect>` comes with a case insensitive term matching, and a small tag indicator to show you how many options were selected, as well a small balloon where you can quickly see what options you have selected so far.

## Full Example

```jsx
const [names, setNames] = useState(initialNames);
const _handleInputChange = value => autocompleteRequest(value); // your autocomplete function

return (
    <FormGroup>
        <FormGroupLabel text="names" />
        <SearchMultiSelect
            placeholder="some placeholder text"
            options={names}
            update={newNames => setNames(newNames)}
            onChange={_handleInputChange}
            context="primary"
            size="sm"
            multiTerm
            searchKeys
            termDivider="-"
        />
    </FormGroup>
);
```

## Options

The format of the options must be an array of objects that contains at least three properties: `name`, `isSelected`, and `key`.

| property   | description                                               |
| ---------- | --------------------------------------------------------- |
| name       | the text displayed on the dropdown                        |
| isSelected | boolean that determines if the option is selected or not  |
| key        | uniquely identifies each options (key must be \`unique\`) |

```
[
    {
        name: 'Jose Santos',
        isSelected: true,
        key: 0
    },
    ...
]
```

## Keybindings

| key   | description                              |
| ----- | ---------------------------------------- |
| enter | toggles the currently highlighted option |
| up    | highlights the previous options          |
| down  | highlights the next option               |
| esc   | closes the dropdown                      |

## Component Example

<Canvas of={SearchMultiSelectStories.Default} />

## Component Example Source

<Source of={SearchMultiSelectStories.Default} />

## SearchMultiSelect Props

<ArgsTable of={SearchMultiSelect} />
