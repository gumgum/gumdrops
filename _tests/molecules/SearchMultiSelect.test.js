/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';

import SearchMultiSelect from '../../components/molecules/SearchMultiSelect';

const defaultProps = {
    options: [
        {
            name: 'Name 1',
            key: 0,
            isSelected: true
        },
        {
            name: 'Name 2',
            key: 1,
            isSelected: false
        }
    ],
    onChange: jest.fn(),
    update: jest.fn(),
    placeholder: 'placeholder',
    size: 'sm',
    className: 'custom-class',
    context: 'primary',
    searchKeys: false,
    multiTerm: false
};

// Note - We cannot write a snapshot test since SearchMultiSelect uses refs which need a DOM
test('Expect <SearchMultiSelect> to show correct options and checked values', () => {
    const { getAllByRole } = render(<SearchMultiSelect {...defaultProps} />);
    const checkboxes = getAllByRole('checkbox');
    expect(checkboxes[0].checked).toEqual(true);
    expect(checkboxes[1].checked).toEqual(false);
});
