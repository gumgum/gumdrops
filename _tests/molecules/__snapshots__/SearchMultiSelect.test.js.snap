// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SearchMultiSelect> to render 1`] = `
<SearchMultiSelect
  className="custom-class"
  context="primary"
  filter={[Function]}
  multiTerm={false}
  onChange={[Function]}
  options={
    Array [
      Object {
        "isSelected": true,
        "key": 10,
        "name": "Name",
      },
    ]
  }
  placeholder="placeholder"
  searchKeys={false}
  size="sm"
  termDivider={/\\[ ,\\]\\+/}
  update={[Function]}
>
  <div
    className="gds-search-select"
  >
    <div
      className="gds-search-select__control"
    >
      <Tag
        className="gds-search-select__tag-indicator gds-search-select__tag-indicator--sm"
        context="primary"
        hasOption={true}
        onClick={[Function]}
        onOptionClick={[Function]}
        optionIcon="bt-times"
        optionLabel="Clear all tags"
        size="xs"
        style={Object {}}
        text="1 Selected"
      >
        <div
          className="gds-tag gds-search-select__tag-indicator gds-search-select__tag-indicator--sm gds-tag--primary gds-tag--with-button gds-tag--with-button--xs gds-tag--xs"
          onClick={[Function]}
          style={Object {}}
        >
          1 Selected
          <button
            aria-label="Clear all tags"
            className="gds-tag__option gds-tag__option--primary gds-tag__option--xs"
            onClick={[Function]}
          >
            <i
              className="btl bt-fw bt-times"
            />
          </button>
        </div>
      </Tag>
      <input
        className="gds-search-select__input gds-search-select__input--sm gds-search-select__input--has-tag"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        placeholder="placeholder"
        type="text"
      />
      <button
        className="gds-search-select__toggle-button -cursor--pointer"
        onClick={[Function]}
        type="button"
      />
    </div>
    <div
      className="gds-search-select__tag-holder
    gds-search-select__tag-holder--bubble "
    >
      <div
        className="gds-search-select__tag-overflow"
      >
        <Tag
          className="-m-a-1"
          context="primary"
          hasOption={true}
          key="10"
          onOptionClick={[Function]}
          optionIcon="bt-times"
          optionLabel="Remove option"
          size="sm"
          style={Object {}}
          text="Name"
        >
          <div
            className="gds-tag -m-a-1 gds-tag--primary gds-tag--with-button gds-tag--with-button--sm gds-tag--sm"
            style={Object {}}
          >
            Name
            <button
              aria-label="Remove option"
              className="gds-tag__option gds-tag__option--primary gds-tag__option--sm"
              onClick={[Function]}
            >
              <i
                className="btl bt-fw bt-times"
              />
            </button>
          </div>
        </Tag>
      </div>
    </div>
    <div
      className="gds-search-select__menu"
    >
      <div
        className="gds-search-select__menu-items"
      >
        <div
          className="gds-search-select__menu-item gds-search-select__menu-item--selected"
          key="10"
          onClick={[Function]}
        >
          <label
            className="gds-search-select__checkbox"
          >
            <input
              checked={true}
              className="gds-search-select__checkbox-input"
              name="Name"
              readOnly={true}
              type="checkbox"
            />
            <span
              className="gds-search-select__checkbox-indicator"
            />
          </label>
          Name
        </div>
      </div>
    </div>
  </div>
</SearchMultiSelect>
`;
